semanticAnalyser/semanticAnalysisVisitor.go

```diff
@@ -0,0 +1,274 @@
+package semanticAnalyser
+
+import "github.com/carlcui/expressive/ast"
+import "github.com/carlcui/expressive/symbolTable"
+import "github.com/carlcui/expressive/typing"
+import "github.com/carlcui/expressive/token"
+import "github.com/carlcui/expressive/logger"
+import "github.com/carlcui/expressive/signature"
+import "fmt"
+
+// SemanticAnalysisVisitor is the general semantic analyser using visitor pattern
+type SemanticAnalysisVisitor struct {
+	logger logger.Logger
+}
+
+// VisitEnterProgramNode creates program scope
+func (visitor *SemanticAnalysisVisitor) VisitEnterProgramNode(node *ast.ProgramNode) {
+	scope := symbolTable.CreateScope(nil)
+	node.SetScope(scope)
+}
+
+// VisitLeaveProgramNode closes program scope
+func (visitor *SemanticAnalysisVisitor) VisitLeaveProgramNode(node *ast.ProgramNode) {
+
+}
+
+// stmts
+
+// VisitEnterVariableDeclarationNode do something
+func (visitor *SemanticAnalysisVisitor) VisitEnterVariableDeclarationNode(node *ast.VariableDeclarationNode) {
+
+}
+
+// VisitLeaveVariableDeclarationNode do something
+func (visitor *SemanticAnalysisVisitor) VisitLeaveVariableDeclarationNode(node *ast.VariableDeclarationNode) {
+
+	identifier := node.Identifier.(*ast.IdentifierNode)
+
+	var resolvedTyping typing.Typing
+
+	if node.DeclaredType == nil && node.Expr == nil {
+		node.SetTyping(typing.ERROR_TYPE)
+		visitor.log(node.GetLocation(), "Missing variable type")
+		return
+	}
+
+	if node.DeclaredType != nil {
+		declaredTyping := node.DeclaredType.GetTyping()
+
+		if node.Expr == nil {
+			resolvedTyping = declaredTyping
+		} else {
+			exprTyping := node.Expr.GetTyping()
+
+			if !exprTyping.Equals(declaredTyping) {
+				visitor.log(node.GetLocation(),
+					"variable declared as "+declaredTyping.String()+", "+
+						"but expression evaluated to "+exprTyping.String())
+
+				resolvedTyping = typing.ERROR_TYPE
+			} else {
+				resolvedTyping = exprTyping
+			}
+		}
+	} else {
+		resolvedTyping = node.Expr.GetTyping()
+	}
+
+	scope := node.GetLocalScope()
+
+	if scope.VariableDeclared(identifier.Tok.Raw) {
+		node.SetTyping(typing.ERROR_TYPE)
+		visitor.log(identifier.GetLocation(), "variable \""+identifier.Tok.Raw+"\" has already been declared")
+		return
+	}
+
+	binding := scope.CreateBinding(identifier.Tok.Raw, identifier.Tok.Locator, resolvedTyping)
+	identifier.SetTyping(resolvedTyping)
+	identifier.SetBinding(binding)
+
+	if node.Tok.TokenType == token.CONST {
+		binding.IsVariable = false
+	}
+
+	node.SetTyping(typing.VOID)
+}
+
+// VisitEnterAssignmentNode do something
+func (visitor *SemanticAnalysisVisitor) VisitEnterAssignmentNode(node *ast.AssignmentNode) {
+}
+
+// VisitLeaveAssignmentNode do something
+func (visitor *SemanticAnalysisVisitor) VisitLeaveAssignmentNode(node *ast.AssignmentNode) {
+
+	identifier := node.Identifier.(*ast.IdentifierNode)
+
+	binding := identifier.GetBinding()
+
+	if binding == nil {
+		node.SetTyping(typing.ERROR_TYPE)
```

Comment:
Maybe put a comment here explaining why binding could be nil and why the error is ignored.

Reply:
"Reason: If the binding is `nil`, then during `visitIdentifierNode` there must be something wrong, and error is handled there. If we pass a null instance (not `nil`) of binding and do normal signature matching, then it will report another error, which is not desirable (and misleading).